velox/exec/OutputBuffer.h

```diff
@@ -124,10 +157,45 @@ class DestinationBuffer {
   // The sequence number of the first item to pass to 'notify'.
   int64_t notifySequence_{0};
   uint64_t notifyMaxBytes_{0};
+  BufferStats stats_;
 };
 
 class Task;
 
+/// The stats of an OutputBuffer.
+struct OutputBufferStats {
+  /// BufferState indicates the OutputBuffer's current state.
+  enum class BufferState {
+    /// The OutputBuffer is open and ready to receive data. Initial state.
+    kOpen,
+    /// The OutputBuffer would not accept more buffers.
+    kNoMoreBuffers,
+    /// The OutputBuffer has received all input data, and some data is still
+    /// to be sent.
+    kFlushing,
+    /// The OutputBuffer has already received and sent all the input data.
+    kFinished
+  };
+
+  OutputBufferStats(
+      core::PartitionedOutputNode::Kind k,
+      bool noMoreBuffers,
```

Comment:
noMoreBuffers vs. canAddBuffers

atEnd vs. canAddPages

Do we need canAddBuffers and canAddPages? Is state not sufficient?

Reply:
After some consideration, I decide to remove the `BufferState` and keep these flags.

 I think the state flags including `canAddBuffers`, `canAddPages` and `finished` are more descriptive than the previous `BufferState` as these they would describe every aspects of the OutputBuffer's status rather than fusing the states into a unified BufferState.
Besides, different engines like presto, spark might have different BufferState definitions, and they could use the state flags to translate to their own version of BufferState.